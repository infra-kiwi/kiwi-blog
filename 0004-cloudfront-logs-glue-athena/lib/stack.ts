import * as cdk from 'aws-cdk-lib';
import { aws_cloudfront, aws_s3, CfnOutput, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import * as aws_glue_alpha from '@aws-cdk/aws-glue-alpha';

export class CloudFrontStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a dedicated S3 bucket to store CloudFront logs
    const logBucket = new aws_s3.Bucket(this, 'CloudFrontLogs', {
      // ACL, required by CloudFront
      objectOwnership: aws_s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
      // Make sure we don't keep unnecessary data after we delete this stack
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true
    });

    const distribution = new aws_cloudfront.Distribution(this, 'Distribution', {
      // HTTPS (because we should ALWAYS use HTTPS!)
      sslSupportMethod: aws_cloudfront.SSLMethod.SNI,

      // Enable logging to the S3 bucket
      enableLogging: true,
      logBucket: logBucket,

      // Use the wonderful https://reqres.in/ as a destination for our calls
      defaultBehavior: {
        origin: new HttpOrigin('reqres.in')
      }
    });
    // This output will contain the domain name of the distribution,
    // e.g. xxxxxxx.cloudfront.net
    new CfnOutput(this, 'DistributionDomainName', {
      value: distribution.domainName
    });

    // Create a database where to store the table
    const database = new aws_glue_alpha.Database(this, 'GlueDatabase', {
      // Give the database a friendly name (lower-case only!)
      databaseName: Stack.of(this).stackName + '-database'
    });
    new CfnOutput(this, 'CloudFrontLogsDatabaseName', {
      value: database.databaseName
    });

    // Create the table that maps each log line
    const table = new aws_glue_alpha.S3Table(this, 'CloudFrontLogsTable', {
      // Give the table a friendly name (lower-case only!)
      tableName: Stack.of(this).stackName + '-cloudfrontlogstable',

      // The source bucket for this Glue table
      bucket: logBucket,
      database: database,

      // Full SQL reference: https://docs.aws.amazon.com/athena/latest/ug/cloudfront-logs.html
      columns: [
        { name: 'date', type: aws_glue_alpha.Schema.DATE },
        { name: 'time', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_location', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'c_ip', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_method', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_host', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_stem', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_status', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'cs_referrer', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_user_agent', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_query', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_cookie', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_request_id', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_host_header', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'time_taken', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_forwarded_for', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_cipher', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_response_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol_version', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_status', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_encrypted_fields', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'c_port', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'time_to_first_byte', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_edge_detailed_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_len', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_start', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_end', type: aws_glue_alpha.Schema.BIG_INT }
      ],

      // Define the right settings to parse CloudFront logs
      dataFormat: aws_glue_alpha.DataFormat.TSV,
      parameters: {
        // THESE PARAMETERS ARE VERY IMPORTANT
        'skip.header.line.count': '2',
        'serialization.format': '\t',
        'field.delim': '\t'
      }
    });
    new CfnOutput(this, 'CloudFrontLogsTableName', {
      value: table.tableName
    });
  }
}
