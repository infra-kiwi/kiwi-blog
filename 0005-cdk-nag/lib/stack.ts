/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_kms, aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use the default AWS-provided SNS KMS key
    const awsSNSKey = aws_kms.Alias.fromAliasName(
      this,
      'AWSManagedSNSKey',
      'alias/aws/sns'
    );

    this.topic = new aws_sns.Topic(this, 'MyTopic', {
      masterKey: awsSNSKey
    });
    new CfnOutput(this, 'MyTopicArn', {
      value: this.topic.topicArn
    });
  }
}
