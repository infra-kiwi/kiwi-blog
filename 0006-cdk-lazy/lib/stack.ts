import * as cdk from 'aws-cdk-lib';
import { aws_sns, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MultiLineParameter } from './multi-line-parameter';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const myParameter = new MultiLineParameter(this, 'MyParameter', {
      parameterName: Stack.of(this).stackName + '-my-parameter'
    });

    // Populate the lines of our multi-line parameter
    myParameter.addLine("It's not DNS");
    myParameter.addLine("There's no way it's DNS");
    myParameter.addLine('It was DNS');

    // Add a dynamic line that depends on another CDK resource's value
    const aTopicAboutDNS = new aws_sns.Topic(this, 'ATopicAboutDNS');
    myParameter.addLine(
      `Here is a topic about DNS: ${aTopicAboutDNS.topicArn}`
    );
    this.topic = aTopicAboutDNS;
  }
}
