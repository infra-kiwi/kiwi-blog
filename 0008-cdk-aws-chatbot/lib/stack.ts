/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ChatBotSlackChannel } from './constructs/chatbot-slack-channel';
import { PostToSlack } from './constructs/post-to-slack';
import { Config } from './util/config';

export type ChatBotStackProps = cdk.StackProps & {
  config: Config;
};

export class ChatBotStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ChatBotStackProps) {
    super(scope, id, props);

    const chatBotSlackChannels: Record<string, ChatBotSlackChannel> = {};

    // Create all the possible ChatBot channel configurations. To do so,
    // extract all available Slack workspace/channel pairs, deduplicate them,
    // and create a ChatBotSlackChannel for each
    const slackChannelEntries = Array.from(
      new Set(
        Object.entries(props.config.notificationChannels).map(([, v]) =>
          JSON.stringify([v.slackWorkspaceId, v.slackChannelId])
        )
      )
    ).map((e) => JSON.parse(e)) as [string, string][];

    for (const entry of slackChannelEntries) {
      const [slackWorkspaceId, slackChannelId] = entry;

      // Configure the channel
      const channel = new ChatBotSlackChannel(
        this,
        `ChatBotSlackChannel-${slackWorkspaceId}-${slackChannelId}`,
        {
          slackWorkspaceId,
          slackChannelId
        }
      );
      chatBotSlackChannels[JSON.stringify(entry)] = channel;
    }

    // Create all the PostToSlack instances, which will accept our custom payloads
    // and translate them into ChatBot-formatted ones
    for (const notificationChannelName in props.config.notificationChannels) {
      const config = props.config.notificationChannels[notificationChannelName];

      // Use the previously-generated Slack channel configuration
      const chatBotSlackChannel =
        chatBotSlackChannels[
          JSON.stringify([config.slackWorkspaceId, config.slackChannelId])
        ];

      new PostToSlack(this, `SlackLambda-${notificationChannelName}`, {
        chatbotSNSTopicArn: chatBotSlackChannel.topicArn,
        notificationChannelName
      });
    }
  }
}
