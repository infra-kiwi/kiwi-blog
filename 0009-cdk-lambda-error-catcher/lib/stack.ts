import * as cdk from 'aws-cdk-lib';
import { aws_lambda, CfnOutput, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import path from 'node:path';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { ErrorCatcherLambda } from './constructs/error-catcher-lambda';
import { LoggingFormat } from 'aws-cdk-lib/aws-lambda';

export type KiwiStackProps = cdk.StackProps & {
  // Expects an SNS topic that accepts a {title,body} payload, like the one configured in
  // https://blog.infra.kiwi/aws-cdk-a-slack-notification-pipeline-via-aws-chatbot-0e0c76e7f4c3
  errorCatcherSNSTopicArn?: string;
};

export class KiwiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: KiwiStackProps) {
    super(scope, id, props);

    // Define our error catcher Lambda, which will listen
    // for error messages in logs
    const errorCatcherLambda = new ErrorCatcherLambda(this, 'Catch', {
      snsTopicArn: props.errorCatcherSNSTopicArn
    });

    // Define a test Lambda that will generate errors and log
    // them using the default TEXT format (props not specified)
    const lambdaWithTextLogs = new NodejsFunction(this, 'TestLambdaTextLogs', {
      functionName: Stack.of(this).stackName + 'test-lambda-text',
      entry: path.join(__dirname, 'lambda/test.ts'),
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: ['@aws-sdk/*'],
        minify: false
      },
      // The Lambda entrypoint name
      handler: 'index.handler'
    });
    new CfnOutput(this, 'TestLambdaTextLogsFunctionName', {
      value: lambdaWithTextLogs.functionName
    });

    // Creates a subscription filter for the LogGroup of the lambdaWithTextLogs
    // function and forwards all errors to the errorCatcherLambda
    errorCatcherLambda.registerErrorCatcher(lambdaWithTextLogs);

    // Define a test Lambda that will generate errors and log
    // them using the JSON format
    const lambdaWithJSONLogs = new NodejsFunction(this, 'TestLambdaJSONLogs', {
      functionName: Stack.of(this).stackName + 'test-lambda-json',
      entry: path.join(__dirname, 'lambda/test.ts'),
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: ['@aws-sdk/*'],
        minify: false
      },
      // The Lambda entrypoint name
      handler: 'index.handler',
      // Define the JSON format for logs
      loggingFormat: LoggingFormat.JSON
    });
    new CfnOutput(this, 'TestLambdaJSONLogsFunctionName', {
      value: lambdaWithJSONLogs.functionName
    });

    // Creates a subscription filter for the LogGroup of the lambdaWithJSONLogs
    // function and forwards all errors to the errorCatcherLambda
    errorCatcherLambda.registerErrorCatcher(lambdaWithJSONLogs);
  }
}
