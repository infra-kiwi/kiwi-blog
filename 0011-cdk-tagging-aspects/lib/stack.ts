/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.topic = new aws_sns.Topic(this, 'MyTopic');

    // Define an output for debugging purposes
    new CfnOutput(this, 'MyTopicArnOutput', {
      value: this.topic.topicArn
    });
  }
}
