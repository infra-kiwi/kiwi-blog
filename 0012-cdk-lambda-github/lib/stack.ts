/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from "aws-cdk-lib";
import {
  aws_apigateway,
  aws_lambda,
  aws_lambda_nodejs,
  aws_secretsmanager,
  CfnOutput,
  Duration,
  Stack,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import path from "node:path";
import {
  secretNameGitHubAppId,
  secretNameGitHubAppInstallationId,
  secretNameGitHubPrivateKey,
} from "./lambda/common-secret";

export class KiwiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a GitHub webhook token. This secret will be automatically
    // filled by Secrets Manager with a random string
    const webhookToken = new aws_secretsmanager.Secret(
      this,
      "GitHubWebhookToken",
      {
        secretName: `${Stack.of(this).stackName}-webhook-token`,
      },
    );
    new CfnOutput(this, "GitHubWebHookTokenSecretName", {
      value: webhookToken.secretName,
    });

    // Define the webhook handler lambda function
    const webhookLambda = new aws_lambda_nodejs.NodejsFunction(
      this,
      "GitHubWebHookLambda",
      {
        functionName: `${Stack.of(this).stackName}-webhook-lambda`,
        entry: path.join(__dirname, "lambda/pr-love.ts"),
        runtime: aws_lambda.Runtime.NODEJS_20_X,
        handler: "index.handler",
        bundling: {
          externalModules: ["@aws-sdk/*"],
        },
        environment: {
          GITHUB_WEBHOOK_TOKEN_SECRET_NAME: webhookToken.secretName,
        },
        // When we interact with the GitHub API, be prepared for longer
        // waiting times!
        timeout: Duration.seconds(10),
      },
    );

    // Grant access to all the various secrets
    webhookToken.grantRead(webhookLambda);
    [
      secretNameGitHubAppId,
      secretNameGitHubAppInstallationId,
      secretNameGitHubPrivateKey,
    ].forEach((secretName) =>
      aws_secretsmanager.Secret.fromSecretNameV2(
        this,
        `Secret-${secretName}`,
        secretName,
      ).grantRead(webhookLambda),
    );

    // Create a new API gateway, which will be the entrypoint for our webhooks
    const gateway = new aws_apigateway.RestApi(this, "GitHubWebHookGateway");

    // Create the webhook handler route /webhook
    const webhook = gateway.root
      .addResource("webhook")
      .addMethod("POST", new aws_apigateway.LambdaIntegration(webhookLambda));

    new CfnOutput(this, "GitHubWebHookUrl", {
      value: gateway.urlForPath(webhook.resource.path),
    });
  }
}
