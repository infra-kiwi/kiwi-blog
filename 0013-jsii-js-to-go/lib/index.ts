/*
 * Copyright (c) 2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

export class Constants {
  static readonly name = 'Mario';
}

export interface Home {
  readonly address: string;
}

export class Addressor {
  getFullGreeting(name: string, home: Home) {
    return `Hello ${Constants.name}! You live at ${home.address}.`;
  }
}
