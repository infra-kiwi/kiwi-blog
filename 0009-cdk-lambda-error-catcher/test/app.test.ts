/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { KiwiStack } from '../lib/stack';
import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

test('KiwiStack', async () => {
  const app = new App();

  // Init stack
  const stack = new KiwiStack(app, 'Test', {
    errorCatcherSNSTopicArn:
      'arn:aws:sns:my-region:123456789012:chatbot-my-channel'
  });

  // Prepare for assertions
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
