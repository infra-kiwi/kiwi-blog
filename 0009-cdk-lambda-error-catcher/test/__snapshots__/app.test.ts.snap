// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KiwiStack 1`] = `
{
  "Outputs": {
    "ErrorCatcherLambdaCatchErrorCatcherLambdaFunctionNameFEA549AD": {
      "Value": {
        "Ref": "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A",
      },
    },
    "ErrorCatcherLambdaCatchErrorCatcherLambdaLogGroupName46F9559A": {
      "Value": {
        "Fn::GetAtt": [
          "ErrorCatcherLambdaCatchErrorCatcherLambdaLogRetention92241BA3",
          "LogGroupName",
        ],
      },
    },
    "TestLambdaJSONLogsFunctionName": {
      "Value": {
        "Ref": "TestLambdaJSONLogs89186D26",
      },
    },
    "TestLambdaTextLogsFunctionName": {
      "Value": {
        "Ref": "TestLambdaTextLogsB2B2502E",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A": {
      "DependsOn": [
        "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRoleDefaultPolicy6AFD8BC9",
        "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRole8C378809",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "fb84d24ddff9bbeab34490acb542b6c2887febed40eefa7e769bb71c9632b8f6.zip",
        },
        "Environment": {
          "Variables": {
            "ERRORS_SNS_TOPIC_ARN": "arn:aws:sns:my-region:123456789012:chatbot-my-channel",
          },
        },
        "FunctionName": "Test-ErrorCatcherLambda-Catch",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRole8C378809",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ErrorCatcherLambdaCatchErrorCatcherLambdaLogRetention92241BA3": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A",
              },
            ],
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRole8C378809": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRoleDefaultPolicy6AFD8BC9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": "arn:aws:sns:my-region:123456789012:chatbot-my-channel",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRoleDefaultPolicy6AFD8BC9",
        "Roles": [
          {
            "Ref": "ErrorCatcherLambdaCatchErrorCatcherLambdaServiceRole8C378809",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4e26bf2d0a26f2097fb2b261f22bb51e3f6b4b52635777b1e54edbd8e2d58c35.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TestLambdaJSONLogs89186D26": {
      "DependsOn": [
        "TestLambdaJSONLogsServiceRole988842C6",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a767cc3c3a836ec15d583d373cbb22a52b5d75e6870efa121b9b18adc4bea6b0.zip",
        },
        "FunctionName": "Testtest-lambda-json",
        "Handler": "index.handler",
        "LoggingConfig": {
          "LogFormat": "JSON",
        },
        "Role": {
          "Fn::GetAtt": [
            "TestLambdaJSONLogsServiceRole988842C6",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestLambdaJSONLogsLogRetention4E685B0D": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "TestLambdaJSONLogs89186D26",
              },
            ],
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "TestLambdaJSONLogsServiceRole988842C6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestLambdaJSONLogsTestLambdaJSONLogsLogGroupErrorCatcherC898E084": {
      "DependsOn": [
        "TestLambdaJSONLogsTestLambdaJSONLogsLogGroupErrorCatcherCanInvokeLambdaB779375C",
      ],
      "Properties": {
        "DestinationArn": {
          "Fn::GetAtt": [
            "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A",
            "Arn",
          ],
        },
        "FilterPattern": "{ $.level = "ERROR" }",
        "LogGroupName": {
          "Fn::GetAtt": [
            "TestLambdaJSONLogsLogRetention4E685B0D",
            "LogGroupName",
          ],
        },
      },
      "Type": "AWS::Logs::SubscriptionFilter",
    },
    "TestLambdaJSONLogsTestLambdaJSONLogsLogGroupErrorCatcherCanInvokeLambdaB779375C": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A",
            "Arn",
          ],
        },
        "Principal": "logs.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":log-group:",
              {
                "Fn::GetAtt": [
                  "TestLambdaJSONLogsLogRetention4E685B0D",
                  "LogGroupName",
                ],
              },
              ":*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "TestLambdaTextLogsB2B2502E": {
      "DependsOn": [
        "TestLambdaTextLogsServiceRoleD704F9A8",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "a767cc3c3a836ec15d583d373cbb22a52b5d75e6870efa121b9b18adc4bea6b0.zip",
        },
        "FunctionName": "Testtest-lambda-text",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "TestLambdaTextLogsServiceRoleD704F9A8",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestLambdaTextLogsLogRetentionDA2424DD": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "TestLambdaTextLogsB2B2502E",
              },
            ],
          ],
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "TestLambdaTextLogsServiceRoleD704F9A8": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TestLambdaTextLogsTestLambdaTextLogsLogGroupErrorCatcher16FAC8E8": {
      "DependsOn": [
        "TestLambdaTextLogsTestLambdaTextLogsLogGroupErrorCatcherCanInvokeLambda87C5F430",
      ],
      "Properties": {
        "DestinationArn": {
          "Fn::GetAtt": [
            "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A",
            "Arn",
          ],
        },
        "FilterPattern": "%^[T:Z\\d.-]+\\t[\\w-]+\\tERROR%",
        "LogGroupName": {
          "Fn::GetAtt": [
            "TestLambdaTextLogsLogRetentionDA2424DD",
            "LogGroupName",
          ],
        },
      },
      "Type": "AWS::Logs::SubscriptionFilter",
    },
    "TestLambdaTextLogsTestLambdaTextLogsLogGroupErrorCatcherCanInvokeLambda87C5F430": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ErrorCatcherLambdaCatchErrorCatcherLambdaB80EEA1A",
            "Arn",
          ],
        },
        "Principal": "logs.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":log-group:",
              {
                "Fn::GetAtt": [
                  "TestLambdaTextLogsLogRetentionDA2424DD",
                  "LogGroupName",
                ],
              },
              ":*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
